File: swift/libipc/Sources/LibIPC/Proto/TypedRouteCodec.swift
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2025-2026 natyamatsya contributors
//
// Generic typed wrapper around Route using a pluggable codec.

public final class TypedRouteCodec<T, C: TypedCodec>: @unchecked Sendable where C.Root == T {

    private let route: Route

    // MARK: Connect

    public static func connect(name: String, mode: Mode) async throws(IpcError) -> TypedRouteCodec<T, C> {
        TypedRouteCodec(route: try await Route.connect(name: name, mode: mode))
    }

    public static func connect(prefix: String, name: String, mode: Mode) async throws(IpcError) -> TypedRouteCodec<T, C> {
        TypedRouteCodec(route: try await Route.connect(prefix: prefix, name: name, mode: mode))
    }

    private init(route: Route) { self.route = route }

    // MARK: Properties

    public var name: String   { route.name }
    public var mode: Mode     { route.mode }
    public var valid: Bool    { route.valid }
    public var recvCount: Int { route.recvCount }

    // MARK: Lifecycle

    public func disconnect() { route.disconnect() }

    // MARK: Send

    /// Send a codec-specific builder payload.
    public func send(builder: C.BuilderType, timeout: Duration = .seconds(200)) throws(IpcError) -> Bool {
        try route.send(data: C.encode(builder: builder), timeout: timeout)
    }

    /// Send raw bytes (already encoded payload).
    public func send(data: [UInt8], timeout: Duration = .seconds(200)) throws(IpcError) -> Bool {
        try route.send(data: data, timeout: timeout)
    }

    // MARK: Recv

    /// Receive a typed message. Returns an empty codec message on timeout.
    public func recv(timeout: Duration? = nil) throws(IpcError) -> C.MessageType {
        C.decode(buffer: try route.recv(timeout: timeout))
    }

    /// Try receiving without blocking.
    public func tryRecv() throws(IpcError) -> C.MessageType {
        C.decode(buffer: try route.tryRecv())
    }

    public func waitForRecv(count: Int, timeout: Duration? = nil) throws(IpcError) -> Bool {
        try route.waitForRecv(count: count, timeout: timeout)
    }

    // MARK: Storage

    public func clear() async { await route.clear() }
    public static func clearStorage(name: String) async { await Route.clearStorage(name: name) }
    public static func clearStorage(prefix: String, name: String) async {
        await Route.clearStorage(prefix: prefix, name: name)
    }
}

File: rust/libipc/src/proto/typed_route_codec.rs
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2025-2026 natyamatsya contributors
//
// Generic typed wrapper around Route using a pluggable codec.

use std::io;

use crate::channel::{Mode, Route};

use super::codec::Codec;

pub struct TypedRouteCodec<T, C> {
    rt: Route,
    _marker: std::marker::PhantomData<(T, C)>,
}

impl<T, C> TypedRouteCodec<T, C>
where
    C: Codec<T>,
{
    /// Connect to a named route as sender or receiver.
    pub fn connect(name: &str, mode: Mode) -> io::Result<Self> {
        Ok(Self {
            rt: Route::connect(name, mode)?,
            _marker: std::marker::PhantomData,
        })
    }

    /// Connect with a prefix.
    pub fn connect_with_prefix(prefix: &str, name: &str, mode: Mode) -> io::Result<Self> {
        Ok(Self {
            rt: Route::connect_with_prefix(prefix, name, mode)?,
            _marker: std::marker::PhantomData,
        })
    }

    pub fn disconnect(&mut self) {
        // Drop and replace with a disconnected state is not directly supported;
        // the route disconnects on Drop. Exposed for API parity with C++.
    }

    /// Send a pre-built typed message.
    pub fn send_builder(&mut self, b: &C::Builder, timeout_ms: u64) -> io::Result<bool> {
        self.rt.send(C::encode(b), timeout_ms)
    }

    /// Send raw bytes (already encoded by the caller).
    pub fn send(&mut self, data: &[u8], timeout_ms: u64) -> io::Result<bool> {
        self.rt.send(data, timeout_ms)
    }

    /// Remove all backing storage for a named route.
    pub fn clear_storage(name: &str) {
        Route::clear_storage(name);
    }

    /// Access the underlying raw route.
    pub fn raw(&mut self) -> &mut Route {
        &mut self.rt
    }

    /// Receive a typed message. Returns an empty message on timeout.
    pub fn recv(&mut self, timeout_ms: Option<u64>) -> io::Result<C::Message> {
        let buf = self.rt.recv(timeout_ms)?;
        Ok(C::decode(buf))
    }

    /// Try receiving without blocking.
    pub fn try_recv(&mut self) -> io::Result<C::Message> {
        let buf = self.rt.try_recv()?;
        Ok(C::decode(buf))
    }
}
